src/serac/physics/thermal_conduction_functional.hpp

```diff
@@ -0,0 +1,481 @@
+// Copyright (c) 2019-2021, Lawrence Livermore National Security, LLC and
+// other Serac Project Developers. See the top-level LICENSE file for
+// details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+
+/**
+ * @file thermal_conduction.hpp
+ *
+ * @brief An object containing the solver for a thermal conduction PDE
+ */
+
+#pragma once
+
+#include "mfem.hpp"
+
+#include "serac/physics/base_physics.hpp"
+#include "serac/numerics/odes.hpp"
+#include "serac/numerics/stdfunction_operator.hpp"
+#include "serac/numerics/functional/functional.hpp"
+#include "serac/physics/state/state_manager.hpp"
+
+namespace serac {
+
+struct ConstantIsotropicMaterial {
+  double rho;
+  double cp;
+  double kappa;
+
+  template <typename T1, typename T2>
+  SERAC_HOST_DEVICE T2 thermalFlux([[maybe_unused]] T1& u, [[maybe_unused]] T2& du_dx) const
+  {
+    return kappa * du_dx;
+  }
+};
```

Comment:
I think we should rename this to something like `LinearIsotropicConductor` or put it in a namespace and name it something like `Thermal::LinearIsotropicMaterial` (otherwise `LinearIsotropicMaterial` might be ambiguous for different physics). I'm not sure what the "Constant" implies in this name (properties don't vary with spatial position => "Homogeneous", kappa doesn't change with temperature => "Linear").

I'm more accustomed to calling it "heat flux" than "thermal flux", as far as the response member function name is concerned, but this is mostly aesthetics. Actually, I'm leaning toward defining the q-function as `MaterialType::operator()`, so that we don't have to introduce a bunch of names for the users/developers to keep track of.

e.g. for a thermomechanical material model, all of the following might be reasonable names for the response function:
1. StressAndHeatFlux
2. HeatFluxAndStress
3. ThermalFluxAndStress
4. ThermalFluxAndCauchyStress
5. ...

and whichever one we chose would be arbitrary, and a user would have to look it up, whereas `MaterialType::operator()` is universal.

Reply:
+1 for `operator()` here